### ISSUE TITLE
allow expanding service-cidr by shortening the prefix length

### ISSUE THREAD

**Issue by dshcherb** [charmed-kubernetes/bundle#600]
On kubernetes master service-cidr charm config hard-codes a /24 subnet which only gives 254 usable service IP addresses.

It is highly likely that users will deploy CDK without noticing this option at first and when their cluster grows larger they will hit the problem with not being able to change that.

While changing the prefix completely is an unsafe operation (https://github.com/kubernetes/kubernetes/issues/12289) making a prefix shorter (e.g. from /24 to /16) is fairly safe as no ClusterIP changes will be needed. Given that we are talking about private IP ranges this will likely be a quite popular solution to the problem.

This subnet is hard-coded in the unit-specific .unit-state-db sqlite database files and checked on charm config changes.

It is possible to do the change by hand but the procedure requires deep understanding of how juju works and how reactive charms are structured:

1) per-unit databases need to be changed by hand:

```
# for all kubernetes-master units do

juju ssh kubernetes-master-<unit-number> sudo -i
cd /var/lib/juju/agents/unit-kubernetes-master-<unit-number>/charm
# apt install sqlite3
sqlite3 .unit-state-db

sqlite> PRAGMA table_info(kv);
0|key|text|0||1
1|data|text|0||0

sqlite> select * from kv where key = 'kubernetes-master.service-cidr';
kubernetes-master.service-cidr|"10.152.183.0/24"


# do NOT forget to add double quotes to the updated data as below - units will error out on update-status with this. If you do it, just fix it and run juju resolved kubernetes-master-<unit-number>
sqlite> update kv set data = '10.152.0.0/16' where key = 'kubernetes-master.service-cidr';
sqlite> select * from kv where key = 'kubernetes-master.service-cidr';
kubernetes-master.service-cidr|10.152.0.0/16

# this is the correct way to do it - double quotes are present and the charm successfully deserializes JSON data
sqlite> update kv set data = '"10.152.0.0/16"' where key = 'kubernetes-master.service-cidr';
sqlite> select * from kv where key = 'kubernetes-master.service-cidr';
kubernetes-master.service-cidr|"10.152.0.0/16"

# exit
```

2) modify current kube-apiserver service args because the re-rendering of that is not triggered

`sed -ie 's/10.152.183.0\/24/10.152.0.0\/16/g' /var/snap/kube-apiserver/current/args`

3) `juju config kubernetes-master service-cidr=''10.152.0.0/16"`

4) restart apiserver services `juju run --application kubernetes-master 'sudo systemctl restart snap.kube-apiserver.daemon.service'`

5) remove the kubernetes service from the default namespace via `kubectl delete -n default svc kubernetes` and kube-dns pod `kubectl delete po kube-dns-<id> -n default` (those will be recreated).

The kubernetes service will use the "old" Cluster-IP which is an immutable field while the kubernetes-master charm will request a new certificate with an IP address in a SAN equal to the first address in the larger subnet which will lead to cert verification failures in some pods (kubernetes-dashboard, nginx-ingress-kubernetes-worker-controller)

(see https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/600#issuecomment-401348384)

**Comment by ktsakalozos** [comment]
Hi @dshcherb , 

We have not moved to a /16 because of backwards compatibility reasons (see https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/272) and support for network restricted environments proxies need to be configured with the source IPs (I do not remember the exact details right now).

Thanks

**Comment by Cynerva** [comment]
What happens to the `kubernetes` service IP when the service CIDR is changed? Does it move from `10.152.183.1` to `10.152.0.1`?

If it changes, we'll need to regenerate the master cert to include the new address ([link](https://github.com/kubernetes/kubernetes/blob/ee2e11a0d43c1534932f00b4dac233369d64f522/cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py#L640)). I would also be worried that existing pods talking to the `kubernetes` service might not pick up the change.

Regardless, I think if the operator really wants to make this change, there should be an easier way to force it.

**Comment by dshcherb** [comment]
@Cynerva 

Replacing a prefix completely is not a good idea as ClusterIPs should remain the same (there is some info that gets written into etcd as well).

Based on my experiment, making a prefix shorter (in other words, the subnet larger) by changing it from /24 to /16 without modifying any bits in the first 16 bits of the old prefix does not trigger any changes in K8s itself and services remain in a good state retaining their ClusterIPs. Any new services get IPs from the bigger subnet.

@ktsakalozos
http proxy stuff will not be relevant as of Juju 2.4 with new juju-http-proxy, juju-https-proxy, juju-no-proxy model-configs which do not modify hook environment if that's the problem you are referring to. I am not even sure it was relevant before because I am not sure charms connect to cluster IPs.

**Comment by ktsakalozos** [comment]
I vaguely remember we had to place the full subnet in a proxy setting (perhaps for node to node communication?) and the /16 was too long.

**Comment by dshcherb** [comment]
@Cynerva 

You're right about the certificate. The charm has a fairly optimistic way of determining a Cluster IP address for the system (which it puts into the SAN for a certificate stored in /root/cdk/server.crt):

```
def get_kubernetes_service_ip():
    '''Get the IP address for the kubernetes service based on the cidr.'''
    interface = ipaddress.IPv4Interface(service_cidr())
    # Add .1 at the end of the network
    ip = interface.network.network_address + 1
    return ip.exploded
```

```

In [1]: import ipaddress

In [2]: net = ipaddress.IPv4Network('10.152.0.0/16')

In [3]: net.network_address + 1
Out[3]: IPv4Address('10.152.0.1')


```

After changing a subnet to a larger one and doing a config change kubernetes dashboard started to fail on connecting to the apiserver as the cert's SAN got changed by the charm code to the first address in the /16 subnet.

**Comment by willthelaw** [comment]
I'm have this situation.  I guess first thing will be to setup a test env to try expanding it, but is the situation the same these days?  Is there anything that can be done to help the cause?  Or words of wisdom.

Long term we plan to move to a new cluster, but for the short term it'd be great to just change.

Thanks!

**Comment by hyperbolic2346** [comment]
@willthelaw My assumption here is that you will run into the same issues as before. You could get around the sans issue with the extra_sans config option, but that seems an unfortunate hack. The other things mentioned here would still be required as well.

**Comment by phvalguima** [comment]
Hi, I looked into this procedure and it worked fine on my lab (k8s 1.16).
I did some tweaks based on later comments here and another idea I had.
In my view, we can separate applications into 3 groups for this problem: (1) applications (and their pods) that do not reach for k8s apis: those are just there, running something; (2) apps that have service accounts and are able to detect changes on "kubernetes service" endpoint: best example I've found openstack-controller-manager as an example; and finally, (3) apps that have service accounts and will not detect changes: that will be all the examples @dshcherb mentioned, kube-dns, dashboard, etc.

So, the first thing is to resolve the certificate. Following @hyperbolic2346, I used extra_sans and added both old and new kubernetes service IP (e.g. 10.152.182.1 and 10.152.183.1).

Also, Kubernetes will eventually move that service endpoint to the new address, as we can see here: https://github.com/kubernetes/kubernetes/blob/abe6321296123aaba8e83978f7d17951ab1b64fd/pkg/master/controller.go#L295

To avoid a situation where k8s randomly move it after this procedure is over, I chose to remove the old (10.152.183.1) kubernetes service endpoint and restarting kube-apiserver. That will be fine for apps on both cases (1) and (2).

Then, I need to account for apps on case (3). One way would be to remove them, either entirely or only their pods, and let them find again the new endpoint as they redeploy. An alternative I tried was to create a second "kubernetes service" which holds the old kubernetes ClusterIP (10.152.183.1). That is just like original service: a service connected to an endpoint that will send traffic to the kube-apiservers. In my case, I can do it with the following yaml:

`
apiVersion: v1           
kind: Service            
metadata:                
  labels:                
    component: apiserver 
    provider: kubernetes 
  name: kubernetes-old 
  namespace: default     
spec:                    
  clusterIP: 10.152.183.1
  ports:                 
  - name: https          
    port: 443            
    protocol: TCP        
    targetPort: 6443     
  type: ClusterIP        
                         
---                      
apiVersion: v1           
kind: Endpoints          
metadata:                
  name: kubernetes-old 
  namespace: default     
subsets:                 
- addresses:             
  - ip: 10.5.0.5         
  - ip: 10.5.0.7         
  ports:                 
  - name: https          
    port: 6443           
    protocol: TCP
`
